Exercise 2

Here We have to locate 12 factories into 12 locations, such that each location gets one factory. 
Let $x_{ij}$ is the variable representing assigning or non assigning of location j to factory i .

\begin{equation*}
x_{ij}=\begin{cases}
          1 \quad &\text{if} \, \text{ factory i is located at jth location} \\
          0 \quad & \, \text{otherwise} \\
     \end{cases}
\end{equation*}
The costs of setting up factory i at location j are given in the txt file.

If $C_{ij}$ is the cost of setting up factory i at location j , then the objective function will be:

$$\ Min \sum_{i=1}^{12}\sum_{j=1}^{12}C_{ij}.x_{ij}$$
s.t.

$$\sum_{j=1}^{12}x_{ij}=1 \ for\ each\ i \in \{1,2,...,12\}$$


$$\sum_{i=1}^{12}x_{ij} = 1 \ for\ each \ j \in \{1,2,...,12\} $$

In [1]:
!pip install -q pyomo

     |████████████████████████████████| 9.1 MB 5.1 MB/s 
     |████████████████████████████████| 49 kB 3.6 MB/s 


In [2]:
from pyomo.environ import * 

In [3]:
import numpy as np

In [4]:
coef = np.loadtxt('lab5_ex2.txt', delimiter='   ')

In [5]:
print('No. of rows: ',coef.shape[0])
print('No. of Columns: ',coef.shape[1])

No. of rows:  12
No. of Columns:  12


In [6]:
print(coef)

[[22. 22. 19. 22. 22. 24. 18. 17. 21. 23. 19. 21.]
 [18. 12. 18. 19. 22. 21. 17. 20. 16. 17. 17. 24.]
 [19. 21. 17. 19. 17. 19. 24. 16. 18. 16. 20. 24.]
 [18. 24. 20. 21. 21. 21. 22. 19. 18. 21. 23. 23.]
 [23. 20. 19. 18. 20. 23. 22. 25. 24. 19. 21. 23.]
 [24. 16. 15. 16. 24. 21. 23. 21. 20. 20. 22. 19.]
 [22. 22. 21. 23. 18. 17. 16. 19. 24. 20. 20. 23.]
 [23. 17. 17. 17. 22. 24. 23. 20. 22. 19. 23. 21.]
 [21. 22. 21. 24. 20. 23. 19. 18. 23. 15. 25. 21.]
 [19. 19. 18. 19. 26. 24. 19. 17. 18. 20. 21. 20.]
 [20. 22. 23. 20. 22. 20. 20. 19. 17. 19. 20. 24.]
 [21. 25. 21. 19. 21. 19. 18. 16. 22. 24. 25. 24.]]


In [7]:
model = ConcreteModel()

In [8]:
N = coef.shape[1]
M = coef.shape[0]
row_indices = np.arange(M)
col_indices = np.arange(N)

In [9]:
model.x = Var(row_indices,col_indices,domain = Binary)

In [10]:
model.constraints = ConstraintList()

In [11]:
model.obj = Objective(expr = sum(coef[i,j]*model.x[i,j] for j in col_indices for i in row_indices),sense = minimize)

In [12]:
for i in row_indices:
  model.constraints.add(sum(model.x[i,j] for j in col_indices)==1)
for j in col_indices:
  model.constraints.add(sum(model.x[i,j] for i in row_indices)==1)


In [13]:
model.pprint()

4 Set Declarations
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :   24 : {1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24}
    x_index : Size=1, Index=None, Ordered=False
        Key  : Dimen : Domain              : Size : Members
        None :     2 : x_index_0*x_index_1 :  144 : {(0, 0), (0, 1), (0, 2), (0, 3), (0, 4), (0, 5), (0, 6), (0, 7), (0, 8), (0, 9), (0, 10), (0, 11), (1, 0), (1, 1), (1, 2), (1, 3), (1, 4), (1, 5), (1, 6), (1, 7), (1, 8), (1, 9), (1, 10), (1, 11), (2, 0), (2, 1), (2, 2), (2, 3), (2, 4), (2, 5), (2, 6), (2, 7), (2, 8), (2, 9), (2, 10), (2, 11), (3, 0), (3, 1), (3, 2), (3, 3), (3, 4), (3, 5), (3, 6), (3, 7), (3, 8), (3, 9), (3, 10), (3, 11), (4, 0), (4, 1), (4, 2), (4, 3), (4, 4), (4, 5), (4, 6), (4, 7), (4, 8), (4, 9), (4, 10), (4, 11), (5, 0), (5, 1), (5, 2), (5, 3), (5, 4), (5, 5), (5, 6), (5, 7), (5, 8), (5, 9), (5, 10), (5, 1

In [14]:
!apt-get install -y -qq coinor-cbc

Selecting previously unselected package coinor-libcoinutils3v5.
(Reading database ... 155047 files and directories currently installed.)
Preparing to unpack .../0-coinor-libcoinutils3v5_2.10.14+repack1-1_amd64.deb ...
Unpacking coinor-libcoinutils3v5 (2.10.14+repack1-1) ...
Selecting previously unselected package coinor-libosi1v5.
Preparing to unpack .../1-coinor-libosi1v5_0.107.9+repack1-1_amd64.deb ...
Unpacking coinor-libosi1v5 (0.107.9+repack1-1) ...
Selecting previously unselected package coinor-libclp1.
Preparing to unpack .../2-coinor-libclp1_1.16.11+repack1-1_amd64.deb ...
Unpacking coinor-libclp1 (1.16.11+repack1-1) ...
Selecting previously unselected package coinor-libcgl1.
Preparing to unpack .../3-coinor-libcgl1_0.59.10+repack1-1_amd64.deb ...
Unpacking coinor-libcgl1 (0.59.10+repack1-1) ...
Selecting previously unselected package coinor-libcbc3.
Preparing to unpack .../4-coinor-libcbc3_2.9.9+repack1-1_amd64.deb ...
Unpacking coinor-libcbc3 (2.9.9+repack1-1) ...
Selecting p

In [15]:
opt_cbc = SolverFactory('cbc')

In [16]:
result = opt_cbc.solve(model)
print('Solver status:', result.solver.status)
print('Solver termination condition:',result.solver.termination_condition)

Solver status: ok
Solver termination condition: optimal


In [17]:
# displaying the solution:
print('Objective Cost of setting up the factories = ', model.obj())

Objective Cost of setting up the factories =  203.0


In [20]:
for i in row_indices:
  for j in col_indices:
    if model.x[i,j].value > 0:
      print('The factory ',i+1,'is located at location ',j+1)
    else:
      pass

The factory  1 is located at location  7
The factory  2 is located at location  2
The factory  3 is located at location  5
The factory  4 is located at location  1
The factory  5 is located at location  11
The factory  6 is located at location  3
The factory  7 is located at location  6
The factory  8 is located at location  4
The factory  9 is located at location  10
The factory  10 is located at location  12
The factory  11 is located at location  9
The factory  12 is located at location  8


Q-10. Changing the integer variables to the real domain variables:

In [21]:
for i in row_indices:
  for j in col_indices:
    model.x[i,j].domain = Reals

In [22]:
#defining upper and lower bounds:
lower_bound = 0
upper_bound = 1

In [23]:
#setting lower and upper bounds:
for i in row_indices:
  for j in col_indices:
    model.x[i,j].setlb(lower_bound)
    model.x[i,j].setub(upper_bound)


In [24]:
model.pprint()

4 Set Declarations
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :   24 : {1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24}
    x_index : Size=1, Index=None, Ordered=False
        Key  : Dimen : Domain              : Size : Members
        None :     2 : x_index_0*x_index_1 :  144 : {(0, 0), (0, 1), (0, 2), (0, 3), (0, 4), (0, 5), (0, 6), (0, 7), (0, 8), (0, 9), (0, 10), (0, 11), (1, 0), (1, 1), (1, 2), (1, 3), (1, 4), (1, 5), (1, 6), (1, 7), (1, 8), (1, 9), (1, 10), (1, 11), (2, 0), (2, 1), (2, 2), (2, 3), (2, 4), (2, 5), (2, 6), (2, 7), (2, 8), (2, 9), (2, 10), (2, 11), (3, 0), (3, 1), (3, 2), (3, 3), (3, 4), (3, 5), (3, 6), (3, 7), (3, 8), (3, 9), (3, 10), (3, 11), (4, 0), (4, 1), (4, 2), (4, 3), (4, 4), (4, 5), (4, 6), (4, 7), (4, 8), (4, 9), (4, 10), (4, 11), (5, 0), (5, 1), (5, 2), (5, 3), (5, 4), (5, 5), (5, 6), (5, 7), (5, 8), (5, 9), (5, 10), (5, 1

In [25]:
opt_cbc = SolverFactory('cbc')

In [26]:
result = opt_cbc.solve(model)
print('Solver status:', result.solver.status)
print('Solver termination condition:',result.solver.termination_condition)

Solver status: ok
Solver termination condition: optimal


In [27]:
# display solution
print('Change of variables to continuos variables:')
print('Factories set up Objective cost :', model.obj())

Change of variables to continuos variables:
Factories set up Objective cost : 203.0


In [28]:
for i in row_indices:
  for j in col_indices:
    if model.x[i,j].value > 0:
      print('The factory ',i+1,'is located at location ',j+1)
    else:
      pass

The factory  1 is located at location  7
The factory  2 is located at location  2
The factory  3 is located at location  5
The factory  4 is located at location  1
The factory  5 is located at location  11
The factory  6 is located at location  3
The factory  7 is located at location  6
The factory  8 is located at location  4
The factory  9 is located at location  10
The factory  10 is located at location  12
The factory  11 is located at location  9
The factory  12 is located at location  8


Ex 2.10 \\
Yes,The optimal costs for both problems are same even after changing the domain of the domains of the variables because the problem is to locate one facility at any particular location.

Q-12. Now due to some reason, facility 2 cannot be assigned to location 4, facility 10 cannot be assigned to location 6 and facility 6 cannot be assigned to location 11. 

So if we make change in the pyomo model, we will  just put up the values of variables x[1,3]=0 , x[9,5] =0 , x[5,10]=0 .

But if we go through making changes in the txt file, then we will change the setup costs of "factory 2 to location 4, factory 10 to location 6 , factory 6 to location 11 " to very high than those of the other setup costs , so that solver avoids to set up these factories to those mentioned locations respectively.

We will assign the values zero to x[1,3],x[9,5],x[5,10] .

In [29]:
#changing the domain again to Binary:
for i in row_indices:
  for j in col_indices:
    model.x[i,j].domain = Binary

In [30]:
model.x[1,3].value = 0
model.x[9,5].value = 0
model.x[5,10].value = 0

In [31]:
model.pprint()

4 Set Declarations
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :   24 : {1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24}
    x_index : Size=1, Index=None, Ordered=False
        Key  : Dimen : Domain              : Size : Members
        None :     2 : x_index_0*x_index_1 :  144 : {(0, 0), (0, 1), (0, 2), (0, 3), (0, 4), (0, 5), (0, 6), (0, 7), (0, 8), (0, 9), (0, 10), (0, 11), (1, 0), (1, 1), (1, 2), (1, 3), (1, 4), (1, 5), (1, 6), (1, 7), (1, 8), (1, 9), (1, 10), (1, 11), (2, 0), (2, 1), (2, 2), (2, 3), (2, 4), (2, 5), (2, 6), (2, 7), (2, 8), (2, 9), (2, 10), (2, 11), (3, 0), (3, 1), (3, 2), (3, 3), (3, 4), (3, 5), (3, 6), (3, 7), (3, 8), (3, 9), (3, 10), (3, 11), (4, 0), (4, 1), (4, 2), (4, 3), (4, 4), (4, 5), (4, 6), (4, 7), (4, 8), (4, 9), (4, 10), (4, 11), (5, 0), (5, 1), (5, 2), (5, 3), (5, 4), (5, 5), (5, 6), (5, 7), (5, 8), (5, 9), (5, 10), (5, 1

In [32]:
opt_cbc = SolverFactory('cbc')

In [33]:
result = opt_cbc.solve(model)
print('Solver status:', result.solver.status)
print('Solver termination condition:',result.solver.termination_condition)

Solver status: ok
Solver termination condition: optimal


In [35]:
print('Optimal Cost for setting up: ',model.obj())

Optimal Cost for setting up:  203.0


In [36]:
# display solution
print('Objective Cost of setting up the factories = ', model.obj())

Objective Cost of setting up the factories =  203.0


In [37]:
for i in row_indices:
  for j in col_indices:
    if model.x[i,j].value > 0:
      print('Factory ',i+1,'is located at location ',j+1)
    else:
      pass

Factory  1 is located at location  7
Factory  2 is located at location  2
Factory  3 is located at location  5
Factory  4 is located at location  1
Factory  5 is located at location  11
Factory  6 is located at location  3
Factory  7 is located at location  6
Factory  8 is located at location  4
Factory  9 is located at location  10
Factory  10 is located at location  12
Factory  11 is located at location  9
Factory  12 is located at location  8


we got the same optimal solution as in the case of non restriction upon the variables x[1,3],x[9,5],x[5,10] as these variables were anyway zero in the basic assignment problem,

And by 0 restriction imposed on it, some extra constraints are imposed in the model and as a result optimal cost for the new model should be greater than or equal to optimal solution of the initial model. But here the optimal solution of the initial model is actually a feasible solution for the changed model also.So optimal solution of the initial model  will be optimal for both the initial and new model. So restricting the Three variables x[1,3],x[9,5],x[5,10] to zero didn't make any effect upon the optimal cost and the values of the other variables.